generic3g/actions/AccumulatorAction.F90

```diff
@@ -0,0 +1,201 @@
+
+      integer :: status
+
+      this%update_calculated = .TRUE.
+
+      _RETURN(_SUCCESS)
+
+   end subroutine calculate_result
+
+   subroutine invalidate(this, importState, exportState, clock, rc)
+      class(AccumulatorAction), intent(inout) :: this
+      type(ESMF_State) :: importState
+      type(ESMF_State) :: exportState
+      type(ESMF_Clock) :: clock
+      integer, optional, intent(out) :: rc
+
+      integer :: status
+      type(ESMF_Field) :: import_field
+      
+      this%update_calculated = .FALSE.
+      call get_field(importState, import_field, _RC)
+      call this%accumulate(import_field, _RC)
+      _RETURN(_SUCCESS)
+
+   end subroutine invalidate
+
+   subroutine get_field(state, field, rc)
+      type(ESMF_State), intent(inout) :: state
+      type(ESMF_Field), intent(inout) :: field
+      integer, optional, intent(out) :: rc
+
+      integer :: status
+      integer :: itemCount
+      integer, parameter :: N = 1
+      character(len=ESMF_MAXSTR) :: itemNameList(N)
+      type(ESMF_StateItem_Flag) :: itemTypeList(N)
+
+      call ESMF_StateGet(state, itemCount=itemCount, _RC)
+      _ASSERT(itemCount == N, 'itemCount does not equal the expected value.')
+      call ESMF_StateGet(state, itemNameList=itemNameList, itemTypeList=itemTypeList, _RC)
+      _ASSERT(itemTypeList(N) == ESMF_STATEITEM_FIELD, 'State item is the wrong type.')
+      call ESMF_StateGet(state, itemName=itemNameList(N), field=field, _RC)
+      _RETURN(_SUCCESS)
+
+   end subroutine get_field
+
+   subroutine accumulate(this, update_field, rc)
+      class(AccumulatorAction), intent(inout) :: this
+      type(ESMF_Field), intent(inout) :: update_field
+      integer, optional, intent(out) :: rc
+      
+      integer :: status
+      type(ESMF_TypeKind_Flag) :: tk, tk_field
+
+      call ESMF_FieldGet(this%accumulation_field, typekind=tk, _RC)
+      call ESMF_FieldGet(update_field, typekind=tk_field, _RC)
+      _ASSERT(tk == tk_field, 'Update field must be the same typekind as the accumulation field.')
+      if(tk == ESMF_TYPEKIND_R4) then
+         call this%accumulate_R4(update_field, _RC)
+      else
+         _FAIL('Unsupported typekind value')
+      end if
+
+      _RETURN(_SUCCESS)
+
+   end subroutine accumulate
+
+   subroutine accumulate_R4(this, update_field, rc)
+      class(AccumulatorAction), intent(inout) :: this
+      type(ESMF_Field), intent(inout) :: update_field
+      integer, optional, intent(out) :: rc
+
+      integer :: status
+      real(kind=ESMF_KIND_R4), pointer :: current(:)
+      real(kind=ESMF_KIND_R4), pointer :: latest(:)
+      real(kind=ESMF_KIND_R4) :: undef
+
+      undef = this%UNDEF_VALUE_R4
+      call assign_fptr(this%accumulation_field, current, _RC)
+      call assign_fptr(update_field, latest, _RC)
+      where(current /= undef .and. latest /= undef)
+        current = current + latest
+      elsewhere(latest == undef)
+        current = undef
+      end where
+      _RETURN(_SUCCESS)
+
```

Comment:
THis seems specific to MeanAccumulator.  Why is it here?

Reply:
Min and Max are the exceptions. They are not accumulations at all. Mean and "Plain" Accumulator are identical except for the additional step in the update method where Mean divides by the counter. So, I kept the accumulate method in the base class because it seems likely that someone will want the "Plain" Accumulator in MAPL3 since they wanted it in MAPL2.

Why did they want it in MAPL2? I have no idea. It doesn't seem like a "Plain" Accumulator would be useful. So I think someone is using an ESMF_Field in a way that we did not anticipate.

The Accumulator subclasses are very similar. The exceptions are at the level of operations on individual elements of the field. That's why I initially had a single derived type that used 3(4) constructors.

Reply:
If you really want to push it, it merely seems to indicate that there is another intermediate subclass that has 2 subclasses for total and mean.  But I don't recommend that.